server/ServerFSM.cpp

```diff
@@ -669,6 +677,91 @@ sc::result MPLobby::react(const JoinGame& msg) {
     return discard_event();
 }
 
+sc::result MPLobby::react(const AuthResponse& msg) {
+    TraceLogger(FSM) << "(ServerFSM) MPLobby.AuthResponse";
+    ServerApp& server = Server();
+    const SpeciesManager& sm = GetSpeciesManager();
+    const Message& message = msg.m_message;
+    const PlayerConnectionPtr& player_connection = msg.m_player_connection;
+
+    std::string player_name;
+    std::string auth;
+    ExtractAuthResponseMessageData(message, player_name, auth);
+
+    if (!server.IsAuthSuccessed(player_name, auth)) {
+        // wrong password
+        player_connection->SendMessage(ErrorMessage(str(FlexibleFormat(UserString("ERROR_WRONG_PASSWORD")) % player_name),
+                                                    true));
+        server.Networking().Disconnect(player_connection);
+        return discard_event();
+    }
+
+    // TODO: Merge it with MPLobby::react(const JoinGame& msg)
+    // assign unique player ID to newly connected player
+    int player_id = server.m_networking.NewPlayerID();
+
+    Networking::ClientType client_type = player_connection->GetClientType();
+
+    // establish player with requested client type and acknowldge via connection
+    player_connection->EstablishPlayer(player_id,
+                    player_name,
+                    client_type,
+                    player_connection->ClientVersionString());
+    player_connection->SendMessage(ContentCheckSumMessage());
+    player_connection->SendMessage(JoinAckMessage(player_id));
+
+    // inform player of host
+    player_connection->SendMessage(HostIDMessage(server.m_networking.HostPlayerID()));
+
+    // Inform AI of logging configuration.
+    if (client_type == Networking::CLIENT_TYPE_AI_PLAYER)
+        player_connection->SendMessage(
+            LoggerConfigMessage(Networking::INVALID_PLAYER_ID, LoggerOptionsLabelsAndLevels(LoggerTypes::both)));
+
+    // assign player info from defaults or from connection to lobby data players list
```

Comment:
if this is duplicating another function's code, should probably put it into another and call from both

Reply:
Merged into new function.